SIPS/sip-58.md

```diff
@@ -0,0 +1,96 @@
+
+### Exchanger.settle ###
+
+Add an internal function `_settlementsOwing` that will emit an event `ExchangeEntryReclaim` for each reclaim and `ExchangeEntryRebate` for each rebate when `Exchanger.settle()` is invoked.
+
+`Exchanger.settle()` will use `_settlementsOwing` to calculate the fee reclamation amounts.
+
+**Event**
+
+Emit an event `ExchangeEntryReclaim` or `ExchangeEntryRebate` for each exchangeEntry within `_settlementsOwing` when `Exchanger.settle()` is invoked.
+
+```solidity
+event ExchangeEntryReclaim(address indexed from, bytes32 src, uint amount, bytes32 dest, uint reclaimAmount, uint srcRoundIdAtPeriodEnd, uint destRoundIdAtPeriodEnd, uint timestamp);
+event ExchangeEntryRebate(address indexed from, bytes32 src, uint amount, bytes32 dest, uint rebateAmount, uint srcRoundIdAtPeriodEnd, uint destRoundIdAtPeriodEnd, uint timestamp);
+```
+
+### Exchanger.appendExchange ###
+
+Emit an event `ExchangeEntryAppended` for each exchangeEntry created when a user makes an exchange. Capture details such as roundIdForSrc, roundIdForDest for Dapps to calculate the effectiveValue of the exchange at anytime by querying the onchain data.
+
+```solidity
+event ExchangeEntryAppended(address indexed account, bytes32 src, uint amount, bytes32 dest, uint amountReceived, uint exchangeFeeRate, uint roundIdForSrc, uint roundIdForDest, uint timestamp);
+```
+
+## Rationale
+
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+
+The decision to add an internal function `_settlementsOwing` that will emit individual fee reclaim and rebate events when `Exchanger.settle()` is invoked allows the public view function `settlementOwing(address account, bytes32 currencyKey)` to be kept for users to query the total aggregated reclaim and rebate amounts they have to settle.
+
+`_settlementsOwing` will be used for settlements and emit the individual events persisting them onto the blockchain once the transaction is confirmed.
+
+## Test Cases
+
+<!--Test cases for an implementation are mandatory for SIPs but can be included with the implementation..-->
+
+- When `Exchanger.settle()` is invoked, the `_settlementsOwing` function is invoked and returns (uint reclaimAmount, uint rebateAmount, uint numEntries).
+- When `Exchanger.settle()` is invoked, it emits `ExchangeEntryReclaim` event with for each ExchangeEntry that has a reclaim amount - (`amountReceived > amountShouldHaveReceived`).
+- When `Exchanger.settle()` is invoked, it emits `ExchangeEntryRebate` event with for each ExchangeEntry that has a rebate amount - (`amountShouldHaveReceived > amountReceived`).
```

Comment:
- Can you also add a test that `ExchangeEntryAppended` is emitted on new exchanges?
- And also where is this emitted - on the Synthetix proxy along with the other events, yes?

Reply:
> And also where is this emitted - on the Synthetix proxy along with the other events, yes?

It would cost more gas to emit all these events off the Synthetix proxy but I guess it makes it easier for tracking events on theGraph as `Exchanger` contract would be upgraded in the future.